pkg/server/api/middleware/authorize_any_of.go

```diff
@@ -0,0 +1,67 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/spiffe/spire/pkg/server/api/rpccontext"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// AuthorizeAnyOf combines authorizers where if any authorizer succeeds, then
+// the caller is authorized. Specifically:
+// 1. If any authorizer returns any status code other than OK or
+// PERMISSION_DENIED, the authorization fails.
+// 2. If all authorizers return PERMISSION_DENIED, then authorization
+// fails.
+// 3. Otherwise, if at least one authorizer returns OK, authorization
+// succeeds.
+func AuthorizeAnyOf(authorizers ...Authorizer) Authorizer {
+	names := make([]string, 0, len(authorizers))
+	for _, authorizer := range authorizers {
+		names = append(names, authorizer.Name())
+	}
+
+	return anyOfAuthorizer{
+		names:       names,
+		authorizers: authorizers,
+	}
+}
+
+type anyOfAuthorizer struct {
+	names       []string
+	authorizers []Authorizer
+}
+
+func (a anyOfAuthorizer) Name() string {
+	return fmt.Sprintf("any-of[%s]", strings.Join(a.names, ","))
+}
+
+func (a anyOfAuthorizer) AuthorizeCaller(ctx context.Context) (context.Context, error) {
+	if len(a.authorizers) == 0 {
+		rpccontext.Logger(ctx).Error("Authorization misconfigured; this is a bug")
```

Comment:
I know that we would rarely run into this, but I'm wondering if we could be more specific here? Something that indicates that there are no authorizers. Mostly because I believe that the same message is logged in the Preprocess for a different reason so it could be difficult to debug.

Reply:
Sure; that's a good suggestion. I'll disambiguate the messaging a little.